### PR TITLE
GPII 3548

### PR DESCRIPTION
**Pull request by danailbd**
Known problems:
- `local` value doesn't get updated with language change. This is being fixed by stegru

### CODE REVIEW

File: src/main/qss.js [hunk 1]
```diff
@@ -415,20 +440,81 @@ gpii.app.qssWrapper.updateSettings = function (that, settings, notUndoable) {
...
+
+
+/**
+ * Updates the given language setting so that it contains a proper list of options. The list includes
+ * only languages that are currently installed on the machine.
+ * @param {Object} settingOptions - The QSS settings specific options
+ * @param {Object[]} installedLanguages - The languages that are currently installed in the system
+ * @param {Object} languageSetting - The language setting that is to be populated with language keys and labels
+ */
+gpii.app.qssWrapper.populateLangSettingOptions = function (settingOptions, installedLanguages, languageSetting) {
```
Comment (georgitodorov) [inline src/main/qss.js]: Use "Language" in the name of the function instead of an abbreviation.

File: src/main/qss.js [hunk 2]
```diff
@@ -415,20 +440,81 @@ gpii.app.qssWrapper.updateSettings = function (that, settings, notUndoable) {
...
+ * @param {String} settingOptions.languageOptionLabelTemplate - The template for every language label
+ * @param {Object[]} installedLanguages - A list of all languages metadata
+ * @return {String[]} The list of labels
+ */
+gpii.app.qssWrapper.getLanguageLabels = function (settingOptions, installedLanguages) {
+    return installedLanguages.map(function (languageMetadata) {
+        return gpii.app.qssWrapper.getLanguageLabel(
+            settingOptions.languageOptionLabelTemplate,
+            languageMetadata,
+            languageMetadata.current
```
Comment (georgitodorov) [inline src/main/qss.js]: Is there a need to pass this as a separate argument given that the whole `languageMetadata` object is provided?

File: src/main/qss.js [hunk 3]
```diff
@@ -415,20 +440,81 @@ gpii.app.qssWrapper.updateSettings = function (that, settings, notUndoable) {
...
+ * Generates a language setting option label out of a provided language metadata.
+ * @param {String} languageOptionLabelTemplate - The template for every language label
+ * @param {Object} languageMetadata - The language metadata that includes different labels
+ * for the language
+ * @param {Boolean} isMainLocale - We wouldn't want to display both the native representation
+ * and the main language as they are redundant
+ * @return {String} The desired language label
+ */
+gpii.app.qssWrapper.getLanguageLabel = function (languageOptionLabelTemplate, languageMetadata, isMainLocale) {
+    languageMetadata.local =
```
Comment (georgitodorov) [inline src/main/qss.js]: `languageMetadata.local = languageMetadata.local || languageMetadata.english` for better readability?

File: src/main/qss.js [hunk 4]
```diff
@@ -478,9 +571,22 @@ gpii.app.qssWrapper.loadSettings = function (that, settingOptions, assetsManager
...
+     * Translations
+     */
+    loadedSettings = gpii.app.qssWrapper.applySettingTranslations(settingOptions, messageBundles, loadedSettings);
+
+
     return loadedSettings;
 };
 
+
+gpii.app.qssWrapper.getSetting = function (settings, path) {
```
Comment (georgitodorov) [inline src/main/qss.js:614]: Missing documentation for this function. Can we use `fluid.find_if` instead of the `find` method for arrays?

File: src/main/qss.js [hunk 5]
```diff
@@ -494,8 +600,10 @@ gpii.app.qssWrapper.alterSetting = function (that, updatedSetting, source) {
         return setting.path === updatedSetting.path && !fluid.model.diff(setting, updatedSetting);
     });
 
+    var settingPath = "settings." + settingIndex;
```
Comment (georgitodorov) [inline src/main/qss.js]: This variable name is ambiguous as each settings have paths in their internal representation. Perhaps revert this change.

File: src/main/qss.js [hunk 6]
```diff
@@ -522,6 +630,96 @@ gpii.app.qssWrapper.getButtonPosition = function (qss, buttonElemMetrics) {
...
+ * @param {Object} settingOptions - Specific options for the QSS settings
+ * @param {Object} messageBundles - Messages for the current locale
+ * @param {Object[]} qssSettingControls - The list of QSS settings to be applied translations to
+ * @return {Object[]} The QSS settings with updated translations
+ */
+gpii.app.qssWrapper.applySettingTranslations = function (settingOptions, messageBundles, qssSettingControls) {
+    var qssSettingMessagesGroup = settingOptions.settingMessagesPrefix,
+        qssSettingMessages = messageBundles[qssSettingMessagesGroup];
+
+    console.log("qssWrapper#applySettingTranslations: ", messageBundles);
```
Comment (georgitodorov) [inline src/main/qss.js:704]: I think that there are too many `console.log` statements in our app already. We should eliminate them where possible.

File: siteconfig.json5 [hunk 7]
```diff
@@ -9,7 +9,14 @@
...
         urls: {
             account: "http://morphic.world/account"
-        }
+        },
+
+        // The template that is used for every label of the language setting's options.
+        // The "%" followed by a word specifies a variable and there are three possible variables:
+        // - native - the name of the language in its native form
+        // - local - the name of the language in the currently applied language
+        // - english - the name of the of the language in English
```
Comment (georgitodorov) [inline siteconfig.json5:18]: `english` is not used or am I missing something?

File: src/renderer/common/js/messageBundles.js [hunk 8]
```diff
@@ -55,6 +63,12 @@
         }
     });
 
+    gpii.psp.fetchCurrentLocale = function (that) {
```
Comment (georgitodorov) [inline src/renderer/common/js/messageBundles.js]: Missing documentation.

File: src/shared/messageBundles.js [hunk 9]
```diff
@@ -195,7 +195,11 @@ gpii.app.messageBundles.loadMessageBundles = function (messageBundlesPath) {
  * @param {String} defaultLocale - The default locale.
  */
 gpii.app.messageBundles.updateMessages = function (that, messageBundles, locale, defaultLocale) {
-    var messages = messageBundles[locale];
+    // make sure the locale is in proper state
+    locale = locale ? locale : "";
```
Comment (georgitodorov) [inline src/shared/messageBundles.js]: `locale = locale || ""`?

**Review by georgitodorov — CHANGES_REQUESTED** [review]
Be sure to run `grunt lint` before merging this PR as currently there are some linting errors.

File: siteconfig.json5 (hunk 7 above)
Comment (danailbd) [reply siteconfig.json5:18]: Yes, you're correct. Currently it is not used, but it is provided by the `gpii.windows.language` module so in case it is needed it can be used

File: src/main/qss.js (hunk 6 above)
Comment (danailbd) [reply src/main/qss.js:704]: I was thinking of replacing all loggers with `fluid.log` which has logging levels.
I think that it is useful for debugging purposes (especially when there was a problem of a user's machine) to have more data.
Of course these loggers should be active only in verbose runs.

File: src/main/qss.js (hunk 2 above)
Comment (danailbd) [reply src/main/qss.js]: This probably needs to be changed. In case we have multiple region options for a language, I suppose we wouldn't want to show the native representation of, lets say `English (Canada)` in case the current language is `English (United States)`

I suggest something like: `languageMetadata.code.startsWith(genericLocale)`

**Comment by danailbd** [comment]
Closing this PR in favour of one that includes Electron 3.x (merged with master)
